### DISCUSSION TITLE
How to have attributes (or source code) shown for `TypedDict` classes

### DISCUSSION THREAD

**Discussion by Andrew-S-Rosen** [mkdocstrings/python#132] [Q&A] [answered]
I have several classes that inherit from `TypedDict` and would like their attributes to be shown in the documentation, such as via the "Source code" viewer (or really, via any means possible). However, what I see is the following:

![image](https://github.com/mkdocstrings/python/assets/8674072/c6b78853-b4d6-4533-b87e-564ddf8d6f57)

for a class like follows:

```python
class RunStatistics(TypedDict):
    """Type hint associated with emmet.core.vasp.calculation.RunStatistics."""

    average_memory: float
    max_memory: float
    elapsed_time: float
    system_time: float
    user_time: float
    total_time: float
    cores: int
```

Any suggestions on how to get these type hints visible to the user so they know what it looks like?

[Here](https://quantum-accelerators.github.io/quacc/reference/quacc/recipes/emt/core.html#quacc.recipes.emt.core.relax_job) is a live, practical example. I want to show the user what the return type is (`OptSchema`), but the `TypedDict` attributes are not shown when one clicks on the type hint.

@pawamoy: This seems potentially somewhat related to https://mkdocstrings.github.io/griffe-pydantic/. Is this covered there or in another package? Thanks!

**Comment by pawamoy** [comment, marked as answer]
You can either:

- document these attributes
  ```python
  class RunStatistics(TypedDict):
      """Type hint associated with emmet.core.vasp.calculation.RunStatistics."""
  
      average_memory: float
      """Docs for average memory."""
      max_memory: float
      """Docs for ..."""
      elapsed_time: float
      """Docs for ..."""
      system_time: float
      """Docs for ..."""
      user_time: float
      """Docs for ..."""
      total_time: float
      """Docs for ..."""
      cores: int
      """Docs for ..."""
  ```
- or set [`show_if_no_docstring: true`](https://mkdocstrings.github.io/python/usage/configuration/docstrings/#show_if_no_docstring)
  ```yaml
  # mkdocs.yml
  plugins:
  - mkdocstrings:
      handlers:
        python:
          options:
            show_if_no_docstring: true
  ```
  or
  ```md
  ::: path.to.RunStatistics
      options:
        show_if_no_docstring: true
  ```

**Reply by Andrew-S-Rosen** [reply]
Perfect, thank you!

**Reply by Andrew-S-Rosen** [reply]
@pawamoy: Doing `show_if_no_docstring: true` works beautifully, but do you know if there is a way to hide the `module-attribute`s that appear as a result? I don't want those to be visible. An example of what I would like to remove is [here](https://quantum-accelerators.github.io/quacc/reference/quacc/recipes/mlp/_base.html#quacc.recipes.mlp._base.logger).

**Reply by pawamoy** [reply]
You can hide them with CSS:

```css
.doc-label-module-attribute {
  display: none;
}
```

**Reply by Andrew-S-Rosen** [reply]
Thanks, @pawamoy. That removes the `module-attribute` label itself but not the actual module attribute from the documentation. For instance, if I have a module-level logger (as in the above hyperlink), I don't need to show that to the end-users.

**Reply by pawamoy** [reply]
You can try to exclude it with [filters](https://mkdocstrings.github.io/python/usage/configuration/members/#filters).
